litestar/_openapi/datastructures.py

```diff
@@ -1,39 +1,99 @@
+class SchemaRegistry:
+    def __init__(self) -> None:
+        self._schema_map: dict = {}
+        self._model_name_groups: defaultdict[str, list[RegisteredSchema]] = defaultdict(list)
+
+    def register(
+        self,
+        key: tuple[str, ...],
+        schema: Schema,
+        reference: Reference,
     ) -> None:
-        """Initialize OpenAPIContext.
+        if (registered_schema := self._schema_map.get(key)) is not None:
+            registered_schema.references.append(reference)
+            return
 
-        Args:
-            openapi_config: OpenAPIConfig instance.
-            plugins: OpenAPI plugins.
-            schemas: Mapping of schema names to schema objects that will become the components.schemas section of the
-                OpenAPI schema.
-        """
+        self._schema_map[key] = registered_schema = RegisteredSchema(key, schema, [reference])
+        self._model_name_groups[key[-1]].append(registered_schema)
+
+    @staticmethod
+    def set_reference_paths(name: str, registered_schema: RegisteredSchema) -> None:
+        for reference in registered_schema.references:
+            reference.ref = f"#/components/schemas/{name}"
+
+    @staticmethod
+    def remove_common_prefix(tuples: list[tuple[str, ...]]) -> list[tuple[str, ...]]:
+        def longest_common_prefix(tuples_: list[tuple[str, ...]]) -> tuple[str, ...]:
+            if not tuples_:
+                return ()
+
+            prefix_ = tuples_[0]
+            for t in tuples_:
+                # Compare the current prefix with each tuple and shorten it
+                prefix_ = prefix_[: min(len(prefix_), len(t))]
+                for i in range(len(prefix_)):
+                    if prefix_[i] != t[i]:
+                        prefix_ = prefix_[:i]
+                        break
+            return prefix_
+
+        prefix = longest_common_prefix(tuples)
+        prefix_length = len(prefix)
+        return [t[prefix_length:] for t in tuples]
+
+    def generate_components_schemas(self) -> dict[str, Schema]:
+        components_schemas: dict[str, Schema] = {}
+
+        for name, name_group in self._model_name_groups.items():
+            if len(name_group) == 1:
+                self.set_reference_paths(name, name_group[0])
+                components_schemas[name] = name_group[0].schema
+                continue
+
+            full_keys = [registered_schema.key for registered_schema in name_group]
+            names = ["_".join(k) for k in self.remove_common_prefix(full_keys)]
+            for name_, registered_schema in zip(names, name_group):
+                self.set_reference_paths(name_, registered_schema)
+                components_schemas[name_] = registered_schema.schema
+
+        return components_schemas
+
+
+class OpenAPIContext:
+    def __init__(
+        self,
+        openapi_config: OpenAPIConfig,
+        plugins: Sequence[OpenAPISchemaPluginProtocol],
+        schemas: dict[str, Schema],
```

Comment:
would like to try and replace `schemas` with `schema_registry` altogether, but not sure if possible yet.